### ISSUE TITLE
BootFailed Cannot Connect To Database

### ISSUE THREAD

**Issue by Emetico** [umbraco/Umbraco-CMS#17363]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

7.15.11, 8.18.15, 10.8.7, 13.5.2

### Bug summary

We are experiencing issues with our Umbraco solutions hosted at Curanet, where the applications lose connection to the database after the server (either web or SQL) has been restarted. This occurs because IIS starts before SQL Server is ready, resulting in stale connections that are not refreshed. Consequently, this leads to a `BootFailed` exception when attempting to start the application. The loss of connection is sporadic; it does not happen every time and can affect different solutions at different times. Our Umbraco solutions are running the following versions:

- Umbraco 7.15.11
- Umbraco 8.18.15
- Umbraco 10.8.7
- Umbraco 13.5.2

### Specifics

- **Connection Loss Timing:** The database connection is lost after a server restart, particularly when IIS initializes before SQL Server.
- **BootFailed Exception:** When the connection issue arises, we receive a BootFailed exception, indicating that the application cannot establish a valid connection to the database due to stale connections.
- **Environment:** All solutions are hosted with Curanet, and the problem is sporadic across different Umbraco versions.

### Steps to reproduce

**1. Environment Setup:** 

- Ensure the Umbraco solutions are hosted on a Curanet server with the following versions:
Umbraco 7.15.11
Umbraco 8.18.15
Umbraco 10.8.7
Umbraco 13.5.2

**2. Restart the SQL Server:**

- Restart the SQL Server instance hosting the database for the Umbraco solution.

**3. Restart the Web Server:**

- Restart the web server (IIS) hosting the Umbraco application.

**4. Monitor Startup Sequence:**

- Observe the startup sequence to ensure that the web server (IIS) starts before the SQL Server is fully operational.

**5. Check Application Logs:**

- After the web server starts, monitor the Umbraco application logs for any entries indicating a BootFailed exception or database connection errors.

**6. Verify Connection Status:**

- Attempt to access the Umbraco backoffice or any front-end page.
- Confirm that the application fails to establish a connection to the database, resulting in a BootFailed error.

### Expected result / actual result

**Expected Result:**

- Upon restarting both the SQL Server and the web server (IIS), the Umbraco application should successfully connect to the database without any errors.
- The application should initialize correctly, allowing access to the backoffice and front-end without encountering a BootFailed exception.

**Actual Result:**

- After restarting the SQL Server and web server, the Umbraco application fails to connect to the database due to stale connections.
- The application logs show a BootFailed exception, indicating that the application could not establish a valid connection to the database.
- Users are unable to access the backoffice or any front-end pages, resulting in service disruption.

**Proposed Solution:**

To address the database connection issues leading to the BootFailed exception, we suggest the following enhancements for Umbraco:

**1. Development of BootFailedNotification:**

- Implement a public UmbracoBootFailedNotification(BootFailedReason reason) as outlined in the [Umbraco notifications documentation](https://docs.umbraco.com/umbraco-cms/reference/notifications).
- This would allow users to check if the reason for the boot failure is related to database connectivity (e.g., CannotConnectToDatabase).
- By detecting this specific failure, applications could use IHostApplicationLifetime.StopApplication() to gracefully stop the application until the database connection is restored.

**2. Utilization of ClearPool():**

- Consider incorporating ClearPool() into Umbraco's error handling for database connections if feasible.
- This could help in clearing stale connections and improving the application's ability to recover from temporary connection issues.

**Comment by Hvid** [comment]
In my opinion, the BootFailedException should not be the last stop. Is there a reason why the boot is not retried?

**Comment by Emetico** [comment]
> In my opinion, the BootFailedException should not be the last stop. Is there a reason why the boot is not retried?

From what I can see from the logs files, it tries to boot around 5 times before it dies. But each time it uses a stale connection that are not refreshed.

**Comment by geoffbeaumont** [comment]
@Emetico What log level do you have Umbraco set to? When I see this on Umbraco 13 I don't get anything other than the usual start-up messages in the log (set to log level Error - and this would definitely seem to be a major error!). Just get an application is not available message in browsers.

Nothing else I work with has this issue with permanently failing if the database is unavailable on start-up, and it's a major (and long standing) problem. An unavailable database is a highly predictable event in many hosting environments, especially if everything is being rebooted following a Windows update - there's no way it should put Umbraco into a permanently unavailable state.

**Comment by VerdelDylan** [comment]
We're experiencing the same issue: after Windows server maintenance, our Umbraco sites (across multiple versions) often fail to reconnect to the database if SQL Server isn't fully ready before IIS starts. This leads to a BootFailedException, and the site remains offline until we manually restart the app pool.

Even when the maintenance timing is carefully adjusted, a delay of just a few milliseconds between services can cause the issue — making it very hard to reliably prevent. Our hosting provider is investigating, but also points to how Umbraco handles boot failures.

Are there any plans to improve boot-time resilience?

**Comment by AndyButland** [comment]
To attempt to at least provide the ability to address these issues, I thought it might be good to give more control to developers on what checks Umbraco does to verify the database is available.  So rather than having a hard-coded 5 times check with a 1 second delay, we ship with that, but give developers scope to implement something else if they prefer.  I've looked at this in https://github.com/umbraco/Umbraco-CMS/pull/19848 - comments welcome.

**Comment by geoffbeaumont** [comment]
@AndyButland I think abstracting this so developers can modify the behaviour to suit their hosting environment is a great idea - but I'd strongly push for making the _default_ implementation a great deal more robust than it currently is. I don't know of any other web  application that's as fragile as Umbraco in this respect, and the vast majority of implementations will be using the default. Both the bug itself and lack of concern over it have done more damage to my clients' perception of Umbraco than anything else. Fortunately, their opinions are otherwise high enough that it hasn't been a deal breaker - so far.

**Comment by Nuklon** [comment]
> [@AndyButland](https://github.com/AndyButland) I think abstracting this so developers can modify the behaviour to suit their hosting environment is a great idea - but I'd strongly push for making the _default_ implementation a great deal more robust than it currently is. I don't know of any other web application that's as fragile as Umbraco in this respect, and the vast majority of implementations will be using the default. Both the bug itself and lack of concern over it have done more damage to my clients' perception of Umbraco than anything else. Fortunately, their opinions are otherwise high enough that it hasn't been a deal breaker - so far.

I agree. The currently implementation is useless. I've touched on this in https://github.com/umbraco/Umbraco-CMS/issues/14968). It may take MSSQL far longer to (re)start. The connect timeout listed there doesn't help. Umbraco just gives up when the server starts. I have several things connected to MSSQL and everything works OK (they just wait until the server/data becomes available). Only Umbraco is always problematic. If you're lucky, it occassionally works, but I think 5 out of 6 times I have to restart the apppool.

**Comment by AndyButland** [comment]
Seems to me there's a need to handle the case where you have the database configured incorrectly too though.  Then just continuing to retry wouldn't seem the best experience - would be better to report the problem so you can adjust as necessary.

But - if you are in a stable production environment - with the abstraction I've suggested you could swap out for one that retries for much longer or just forever if preferred.

**Comment by Nuklon** [comment]
> Seems to me there's a need to handle the case where you have the database configured incorrectly too though. Then just continuing to retry wouldn't seem the best experience - would be better to report the problem so you can adjust as necessary.

Except I _know_ it's configured correctly. Almost every Windows Update I have to recycle IIS' AppPool of Umbraco. It's the only app I use that's requires it, so I agree with @geoffbeaumont that the default Umbraco implementation should be more robust than what it is right now.

**Comment by AndyButland** [comment]
> Except I know it's configured correctly. 

Point noted.  I suggest now we have the abstraction in place though, it would be straightforward to implement an alternative that just continues to retry indefinitely if you prefer that, and can sure about the configuration.  Or one with more retries, longer delays, or something more sophisticated like exponential backoff.

Given that I'll close the issue now, as we have a solution that prevents you being limited by the default behaviour of Umbraco.

Would be happy to take contributions of alternative implementations of the abstraction that we could either document, or ship with in core, to allow people to just register rather than implement an alternative considered better for production than the default (which is perhaps more geared to initial setup, local development and less stable environments).

Resolved via https://github.com/umbraco/UmbracoDocs/pull/7286 for 16.2.

Cross-referencing the documentation update that will go live with the 16.2 release candidate: https://github.com/umbraco/UmbracoDocs/pull/7286
